pydantic/fields.py

```diff
@@ -935,6 +935,35 @@ def _validate_singleton(
     ) -> 'ValidateReturn':
         if self.sub_fields:
             errors = []
+
+            if is_union_origin(get_origin(self.type_)) and self.model_config.smart_union:
```

Comment:
```suggestion
            if self.model_config.smart_union and is_union_origin(get_origin(self.type_)):
```
Should be quicker since it's just checking a bool, if that is false we skip the `is_union_origin` check.